Plumbing[Reaction Chamber & Chemical Filter] now pick reagents from a drop down list

SyncIt21 opened the pull request:
## About The Pull Request
Plumbing[Reaction Chamber & Chemical Filter] & other tools such as bloodfilter etc which depend on the tools.dm file now all pick reagents from an drop down list instead of manually typing it in a text field

## Why It's Good For The Game
Entering long reagent names in a text field is tiring and error prone. Selecting the reagent from the drop down list with an search bar attached at the bottom makes this process much faster and reduces typing errors to zero

SyncIt21 committed:
Description:Modifies Plumbing Reaction Chamber & Chemical Filter to choose reagents reagents
from a drop down list instead of manually typing their names in a textfield

Why Its Good : Typing reagent names especially long names is very error prone and tiering
choosing from a drop down list which has an search bar atatched to its bottom
makes this process much faster and less error prone

SyncIt21 committed:
Modifies Plumbing Reaction Chamber & Chemical Filter to choose reagents reagents
from a drop down list instead of manually typing their names in a textfield

Why Its Good : Typing reagent names especially long names is very error prone and tiering
choosing from a drop down list which has an search bar atatched to its bottom
makes this process much faster and less error prone

Side Effect : Blood Filter UI also modified as it inherits from Chem Filter UI

SyncIt21 committed:
Merge branch 'master' of https://github.com/SyncIt21/tgstation

SyncIt21 committed:
Merge branch 'tgstation:master' into master

SyncIt21 committed:
Fixes BloodFilter to use Plumbing ChemFilter UI correctly

Now blood filter also uses a drop down list for picking reagents

SyncIt21 committed:
Merge branch 'master' of https://github.com/SyncIt21/tgstation

SyncIt21 committed:
Prettier code refractoring

the-orange-cow commented:
I haven't touched plumbing in a while, but it's my understanding that you can add chemicals outside of the normally synthesized ones to a plumbing system.  This being the case it stands to reason that the dropdown list would need to include litterally every reagent in the game(to include drinks).

Assuming I'm correct in the above, is that feasable?

SyncIt21 commented:
Including all reagent names was very easy actually. Inside the code/modules/reagents/chemistry/holder.dm file there is a method called init_chemical_reagent_list(). All i did was just extract the names from the returned list and use that list in plumbing

Time-Green commented:
Could you show a very small video or a picture of you using the UI?

SyncIt21 commented:
**1) Blood Filter UI**

a) Picking reagent to use in blood filter
![Blood Filter Reagent List](https://user-images.githubusercontent.com/110812394/199701464-bb0e6ea3-f04f-4fa6-8574-ce497212f14b.png)

b)Blood reagent added
![Blood Filter Reagent Added](https://user-images.githubusercontent.com/110812394/199701530-41c24e2b-a39c-4e70-b4a0-158fe037087f.png)


**2) Plumbing reaction chamber UI**

a) Picking reagents
![Plumbing Reaction Chamber Drop Down List](https://user-images.githubusercontent.com/110812394/199701690-1aed1b29-0af8-4805-b25b-f159cdde2c7b.png)

b) Reagents added
![Plumbing Reaction Chamber Reagent Added](https://user-images.githubusercontent.com/110812394/199701744-4d360a0b-9ccc-4be9-8ac6-fe709213b0e1.png)


**3) Chem filter UI**

a) Adding left reagent
![Add left reagent to chem filter](https://user-images.githubusercontent.com/110812394/199701873-fc3456f5-0054-4dfc-87b7-710254cbbd94.png)

b) Adding right reagent
![Adding right reagent to chem filter](https://user-images.githubusercontent.com/110812394/199702030-ffbe1979-9be1-41f8-8cf6-0a00dddc00d6.png)

c) Both reagents added
![Chm Filter Left   Right Reagent Added](https://user-images.githubusercontent.com/110812394/199702084-000c7687-4611-4606-b783-4425c06a4068.png)

SyncIt21 committed:
PR

SyncIt21 committed:
OOOOFFFF

OOOOFFFF

SyncIt21 committed:
Merge branch 'RPED&BRPED' of https://github.com/SyncIt21/tgstation

SyncIt21 committed:
X

Mothblocks commented on .vscode/settings.json:
```diff
@@ -43,5 +43,8 @@
 	},
-	"Lua.diagnostics.enable": false
+	"Lua.diagnostics.enable": false,
+	"githubPullRequests.ignoredPullRequestBranches": [
```
What is this change?

SyncIt21 replied:
Again I'm sorry still new to github.  Didn't want to include this I have no idea how to uncommit this. U can remove this change

Mothblocks commented on .vscode/tasks.json:
```diff
@@ -74,7 +74,10 @@
 				"$eslint-stylish"
 			],
-			"group": "build",
+			"group": {
```
What is this change?

SyncIt21 replied:
Ok these are just unnecessary . Committed them by mistake you can remove them

Mothblocks replied:
Can you please do it

Mothblocks commented on code/_globalvars/lists/objects.dm:
```diff
@@ -20,6 +20,7 @@ GLOBAL_LIST(chemical_reactions_list) //list of all /datum/chemical_reaction datu
 GLOBAL_LIST(chemical_reactions_list_reactant_index) //list of all /datum/chemical_reaction datums. Used during chemical reactions. Indexed by REACTANT types
 GLOBAL_LIST(chemical_reactions_list_product_index) //list of all /datum/chemical_reaction datums. Used for the reaction lookup UI. Indexed by PRODUCT type
 GLOBAL_LIST_INIT(chemical_reagents_list, init_chemical_reagent_list()) //list of all /datum/reagent datums indexed by reagent id. Used by chemistry stuff
+GLOBAL_LIST_INIT(chemical_name_list, init_chemical_name_list()) //names of reagents used by plumbing UI.
```
```suggestion
/// names of reagents used by plumbing UI.
GLOBAL_LIST_INIT(chemical_name_list, init_chemical_name_list())
```

`///` == auto-doc comment. Yes I know the others are wrong

SyncIt21 replied:
Thanks

Mothblocks commented on code/modules/plumbing/plumbers/filter.dm:
```diff
@@ -37,20 +37,22 @@
+								english_right += selected_reagent
+								right += chem_id
+				else
+					to_chat(usr, span_warning("No such Known Reagent Exists!!"))
```
```suggestion
					CRASH("User gave unknown chemical ID: [selected_reagent]")
```

Because this is a `tgui_input_list`, we can't get unsanitized input here, so this would be a failure case.

Mothblocks commented on code/modules/plumbing/plumbers/reaction_chamber.dm:
```diff
@@ -86,18 +87,21 @@
+					var/input_amount = text2num(params["amount"])
+					if(input_amount)
+						required_reagents[input_reagent] = input_amount
+						. = TRUE
```
Remove these and just have `return TRUE` at the very end of the proc itself.

Mothblocks commented on code/modules/plumbing/plumbers/reaction_chamber.dm:
```diff
@@ -86,18 +87,21 @@
 		return
 	switch(action)
+		if("add")
+			var/selected_reagent = tgui_input_list(usr, "Select Reagent", "Reagent", GLOB.chemical_name_list)
```
Do most tgui_input_list do "Select Reagent" instead of "Select reagent"?

SyncIt21 replied:
I'll change it to Select reagent

Mothblocks commented on code/modules/reagents/chemistry/holder.dm:
```diff
@@ -20,6 +20,15 @@
 
+/// Creates an list which is indexed by reagent name . used by plumbing reaction chamber and chemical filter UI
+/proc/init_chemical_name_list()
+	var/list/name_list=new/list()
```
```suggestion
	var/list/name_list = list()
```

Mothblocks commented on code/modules/surgery/tools.dm:
```diff
@@ -524,13 +524,13 @@
+				if(chem_id && !(chem_id in whitelist))
+					whitelist[chem_id] = selected_reagent
+				else
+					to_chat(usr, span_warning("Invalid Reagent Or Already In List!"))
```
CRASH here

SyncIt21 replied:
So u replace the code inside else condition with CRASH()? Sure but I don't think else condition will ever be executed since it will always return a valid name from the list. Maybe I remove the else condition all together?

Mothblocks replied:
It is up to you--it never being executed is why you would want a CRASH there, so we get a stack trace when it happens, but it's not necessary and would only be if you find it useful (or if the invariant is something shakier)

Mothblocks commented on code/modules/surgery/tools.dm:
```diff
@@ -524,13 +524,13 @@
-			if(!(chem_id in whitelist))
-				whitelist[chem_id] = new_chem_name
+			var/selected_reagent = tgui_input_list(usr, "Select Reagent To Filter", "WhiteList Reagent", GLOB.chemical_name_list)
+			if(selected_reagent)
```
Early-return

SyncIt21 replied:
I early return where exactly? U mean if(!selected_reagent)
  return
?

Mothblocks replied:
Yes

Mothblocks commented on tgui/packages/tgui/interfaces/ChemFilter.tsx:
```diff
@@ -11,37 +11,30 @@ type Data = {
 export const ChemFilterPane = (props: Props, context) => {
   const { act } = useBackend(context);
-  const { title, list, reagentName, onReagentInput } = props;
+  const { title, list, btitle, bcolor } = props;
```
Do not abbreviate names. Fully expand this to `buttonTitle` and `buttonColor`.

SyncIt21 replied:
My bad ill fix that

Mothblocks reviewed (changes requested):
Ideally, this would not be a separate tgui_input_list, but instead a part of the UI itself. I'm thinking of just being able to type and see autocomplete results show up in real time. I don't expect you to do that though.

SyncIt21 commented on .gitignore:
```diff
@@ -222,3 +222,4 @@ libprof.so
 
 # Screenshot tests
 /artifacts
+tgstation_unit_test_focus_file.dm
```
My bad this is my very first time using github and vs code I didnt want to include this here.  I'll remove it

SyncIt21 commented on code/modules/plumbing/plumbers/filter.dm:
```diff
@@ -37,20 +37,22 @@
-			else
-				to_chat(usr, span_warning("No such known reagent exists!"))
+			var/which=params["which"]
+			var/selected_reagent = tgui_input_list(usr, "Select [which] Reagent", "Reagent", GLOB.chemical_name_list)
```
True I will change the title inside ChemReactionChamber.tsx file to capital "Left" and "Right" then it will be consistent

SyncIt21 commented on code/modules/reagents/chemistry/holder.dm:
```diff
@@ -20,6 +20,15 @@
+	var/list/name_list=new/list()
+	for(var/X in GLOB.chemical_reagents_list)
+		var/datum/reagent/R = GLOB.chemical_reagents_list[X]
+		name_list[R.name]=R.name
```
Sorry I'm new to using lists.  I didn't know about your suggestion. Do I just do name_list.add()? Or something similar. I come from an java ecosystem so this is all new to me

Mothblocks replied:
`name_list += reagent.name`

SyncIt21 commented on code/modules/surgery/tools.dm:
```diff
@@ -524,13 +524,13 @@
-				return
-			if(!(chem_id in whitelist))
-				whitelist[chem_id] = new_chem_name
+			var/selected_reagent = tgui_input_list(usr, "Select Reagent To Filter", "WhiteList Reagent", GLOB.chemical_name_list)
```
I'll make it all lower case

SyncIt21 committed:
Update code/modules/reagents/chemistry/holder.dm

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 committed:
Update code/modules/plumbing/plumbers/filter.dm

Yup good point

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 committed:
Update code/_globalvars/lists/objects.dm

Good point

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

SyncIt21 committed:
Update tasks.json

Removed unintentional commit

SyncIt21 committed:
Delete .gitignore

Unintentional commit removed

SyncIt21 committed:
Delete settings.json

Unintentional commit removed

SyncIt21 committed:
Delete tasks.json

Unintentional commit removed

SyncIt21 committed:
Update reaction_chamber.dm

Changed Prompt in tgui list from "Select Reagent" to "Select reagent"

SyncIt21 committed:
Update filter.dm

Replaced CRASH with early returns on lines 43 and 47

SyncIt21 committed:
Update reaction_chamber.dm

Added Early returns at lines 93 & 97 and replaced all occurences of . =TRUE with return TRUE at the end of the proc except for adding where we do early return instead

SyncIt21 committed:
Update holder.dm

replaced needless mapping with just normal name add at line 28

SyncIt21 committed:
Update tools.dm

added early returns on lines 529 and 533, removing the need for CRACH or else statements

SyncIt21 committed:
Update tools.dm

removed prompt capitalization on line 527 and added early returns

SyncIt21 committed:
Update ChemFilter.tsx

Added full names buttonTitle & buttonColor

SyncIt21 committed:
Update tools.dm

removed excess chem_id check on line 535

SyncIt21 committed:
Update BloodFilter.tsx

ChemFilterPane uses buttonTitle & buttonColor instead of bTitle & bColor

SyncIt21 committed:
Update filter.dm

Changed prompt from "Select [which] Reagent" to "Select [which] reagent" small 'r' to match case

SyncIt21 reviewed (commented):
1) Early returns added
2) Changed capitalization of text in some prompts
3) removed needless mapping of elements in list

SyncIt21 commented:
> Ideally, this would not be a separate tgui_input_list, but instead a part of the UI itself. I'm thinking of just being able to type and see autocomplete results show up in real time. I don't expect you to do that though.

Thats a good idea but i don't have much experience in UI . Just started contributing yesterday hopefully will make that change once i learn enough

SyncIt21 committed:
Update holder.dm

spelling mistake at line 28. Changed from reagent to R

SyncIt21 committed:
Update ChemFilter.tsx

Forgot to replace all bTitle & bColor occurences. Now fixed

SyncIt21 committed:
RPED & B-RPED can now pick up cable,glass,plasteel and display parts in a more verbose way

Save space inside backpack from these parts and read part info much easily

SyncIt21 committed:
Merge branch 'tgstation:master' into master

SyncIt21 committed:
Merge branch 'tgstation:master' into RPED&BRPED

SyncIt21 committed:
Merge branch 'tgstation:master' into RPED&BRPED

SyncIt21 commented:
I made tbis PR saccidentally on my project main branch as a result  PR [#71064](https://github.com/tgstation/tgstation/pull/71064) which  branches from the main branch should not be merged until this PR is first approved/closed

SyncIt21 committed:
Merge branch 'tgstation:master' into master

SyncIt21 committed:
Merge branch 'tgstation:master' into master

SyncIt21 committed:
Merge branch 'tgstation:master' into RPED&BRPED

SyncIt21 committed:
Merge branch 'RPED&BRPED'

SyncIt21 committed:
Merge branch 'tgstation:master' into master

SyncIt21 committed:
Create .gitignore
